### DISCUSSION TITLE
Integration of Hackieuino improvements

### DISCUSSION THREAD

**Discussion by SciLor** [biologist79/ESPuino#251] [Show and tell]
I have started porting ESPuino to the Toniebox ESP32 version with the esp32-s3.
I am currently merging my changes I made on the master branch to the dev branch.
Beside the Toniebox specific things (I try to implement into special classes), there are several improvements in the code, that may be interesting for the project.
Maybe you could already cherry-pick some commits? Ex:

4-bit SDMMC support, some text literal corrections, refractor the deepsleep into an extra method, implement a restart function, trf7962a rfid reader, lis3dh acceleration sensor, rgb-led support

https://github.com/biologist79/ESPuino/compare/dev...toniebox-reverse-engineering:Hackieuino:dev

**Comment by biologist79** [comment]
Hi,
I really appreciate your work but I don't want to integrate T****box-stuff into ESPuino. Not because I wouldn't like it, but the thing is that there's company (with stake/shareholders) behind it and basically they earn the better part of their money by selling T****s. T****box itself is sorta cheap. By providing an easy way to use their box but circumvent the need to buy T****s, that would be an direct offensive to their business model. And I really want to make sure this doesn't happen as I really don't want to run into lawsuit with them. By the way this is why I never use their trademarks in order to make sure there's not "link" between ESPuino and them.

However, 4bit SDMMC support would be ok for me.

**Reply by SciLor** [reply]
Hey @biologist79.
Thank you for your reply.
 
So, then I would suggest just integrating the hardware peripherals the box uses, as it can be also used in custom hardware
Beside that, keeping the Tb-specific class and configuration out of your repository may be a solution to address your concerns.

PS: We're already circumventing the need for the figurines by providing a tool that allows filling the internal microSD and using custom tags. Fingers crossed, currently nobody started to bother us for that yet.

**Comment by tueddy** [comment]
I have an ESP32-S3 & SD-adapter here, can test & apply your SD-MMC 4Bit code. One question:
Currently we can reach this upload speed (via webinterface):

- SPI: 350KB/s 
- MMC 1Bit: 550KB/s
- MMC 1Bit with "Arduino as component": 700KB/s

4 Bit MMC should be 20% faster than 1Bit according to https://www.instructables.com/Select-SD-Interface-for-ESP32/

How fast is webupload rate with your hardware?

**Reply by SciLor** [reply]
Sorry for my late reply. I am currently focusing on a different project. I hope I can test this, end of next week (as I need to resetup my box for dev again)

**Comment by tueddy** [comment]
I don't own this hardware but made a test with a ESP32 Lolin S3 + SD card (1 & 4 bit mode):

1 Bit MMC:

SD card read: 1048576 bytes read for 316 ms
SD card write: 1048576 bytes written for 1057 ms
web upload: 6346838 bytes in 10026 ms (633 kiB/s)



4 Bit MMC:

SD card read: 1048576 bytes read for 163 ms
SD card write: 1048576 bytes written for 881 ms
web upload:  6346838 bytes in 9555 ms (664 kiB/s)

So 4 Bit MMC is not really faster in webload, maybe due to limited HTTP upload rate..
